zenoh/src/selector.rs

```diff
@@ -66,377 +66,200 @@ use std::{
-        if let Cow::Borrowed(s) = self.parameters {
-            self.parameters = Cow::Owned(s.to_owned())
-        }
-        if let Cow::Owned(s) = &mut self.parameters {
-            s
-        } else {
-            unsafe { std::hint::unreachable_unchecked() } // this is safe because we just replaced the borrowed variant
-        }
+    pub fn parameters_mut(&mut self) -> &mut Parameters<'a> {
+        &mut self.parameters
     }
-    pub fn set_parameters(&mut self, selector: impl Into<Cow<'a, str>>) {
-        self.parameters = selector.into();
-    }
-    pub fn borrowing_clone(&'a self) -> Self {
-        Selector {
-            key_expr: self.key_expr.clone(),
-            parameters: self.parameters.as_ref().into(),
-        }
+
+    /// Sets the `parameters` part of this `Selector`.
+    #[inline(always)]
+    pub fn set_parameters<P>(&mut self, parameters: P)
+    where
+        P: Into<Parameters<'static>>,
+    {
+        self.parameters = parameters.into();
     }
+
+    /// Create an owned version of this selector with `'static` lifetime.
     pub fn into_owned(self) -> Selector<'static> {
         Selector {
             key_expr: self.key_expr.into_owned(),
-            parameters: self.parameters.into_owned().into(),
+            parameters: self.parameters.into_owned(),
         }
     }
 
-    #[deprecated = "If you have ownership of this selector, prefer `Selector::into_owned`"]
-    pub fn to_owned(&self) -> Selector<'static> {
-        self.borrowing_clone().into_owned()
-    }
-
     /// Returns this selectors components as a tuple.
-    pub fn split(self) -> (KeyExpr<'a>, Cow<'a, str>) {
+    pub fn split(self) -> (KeyExpr<'a>, Parameters<'a>) {
         (self.key_expr, self.parameters)
     }
 
-    /// Sets the `parameters` part of this `Selector`.
-    #[inline(always)]
-    pub fn with_parameters(mut self, parameters: &'a str) -> Self {
-        self.parameters = parameters.into();
-        self
+    #[zenoh_macros::unstable]
+    /// Sets the time range targeted by the selector.
+    pub fn set_time_range<T: Into<Option<TimeRange>>>(&mut self, time_range: T) {
+        self.parameters_mut().set_time_range(time_range);
     }
 
-    pub fn extend<'b, I, K, V>(&'b mut self, parameters: I)
-    where
-        I: IntoIterator,
-        I::Item: std::borrow::Borrow<(K, V)>,
-        K: AsRef<str> + 'b,
-        V: AsRef<str> + 'b,
-    {
-        let it = parameters.into_iter();
-        let selector = self.parameters_mut();
-        let mut encoder = form_urlencoded::Serializer::new(selector);
-        encoder.extend_pairs(it).finish();
+    #[zenoh_macros::unstable]
+    /// Extracts the standardized `_time` argument from the selector parameters.
+    ///
+    /// The default implementation still causes a complete pass through the selector parameters to ensure that there are no duplicates of the `_time` key.
+    pub fn time_range(&self) -> ZResult<Option<TimeRange>> {
+        self.parameters().time_range()
```

Comment:
Do we really need these functions now, as the publicly available `.parameters()`/`.parameters_mut()` supports them too? Same about `accept_any_keyexpr`

Reply:
I've marked them as `unstable` since it's unclear yet the impact on the rest. `accept_any_keyexpr` could be removed since it's `pub(crate)`, `time_range()` also performs deserialization of the parameter if present. The impact of removing `time_range()` should be further investigated.